// @effect/vitest/dist/esm/index.js
import * as V from "vitest";
import * as internal from "./internal/internal.js";
/**
 * @since 1.0.0
 */
export * from "vitest";
/**
 * @since 1.0.0
 */
export const addEqualityTesters = internal.addEqualityTesters;
/**
 * @since 1.0.0
 */
export const effect = internal.effect;
/**
 * @since 1.0.0
 */
export const scoped = internal.scoped;
/**
 * @since 1.0.0
 */
export const live = internal.live;
/**
 * @since 1.0.0
 */
export const scopedLive = internal.scopedLive;
/**
 * Share a `Layer` between multiple tests, optionally wrapping
 * the tests in a `describe` block if a name is provided.
 *
 * @since 1.0.0
 *
 * ```ts
 * import { expect, layer } from "@effect/vitest"
 * import { Context, Effect, Layer } from "effect"
 *
 * class Foo extends Context.Tag("Foo")<Foo, "foo">() {
 *   static Live = Layer.succeed(Foo, "foo")
 * }
 *
 * class Bar extends Context.Tag("Bar")<Bar, "bar">() {
 *   static Live = Layer.effect(
 *     Bar,
 *     Effect.map(Foo, () => "bar" as const)
 *   )
 * }
 *
 * layer(Foo.Live)("layer", (it) => {
 *   it.effect("adds context", () =>
 *     Effect.gen(function* () {
 *       const foo = yield* Foo
 *       expect(foo).toEqual("foo")
 *     })
 *   )
 *
 *   it.layer(Bar.Live)("nested", (it) => {
 *     it.effect("adds context", () =>
 *       Effect.gen(function* () {
 *         const foo = yield* Foo
 *         const bar = yield* Bar
 *         expect(foo).toEqual("foo")
 *         expect(bar).toEqual("bar")
 *       })
 *     )
 *   })
 * })
 * ```
 */
export const layer = internal.layer;
/**
 * @since 1.0.0
 */
export const flakyTest = internal.flakyTest;
/**
 * @since 1.0.0
 */
export const prop = internal.prop;
/**
 * @since 1.0.0
 */
/** @ignored */
const methods = {
  effect,
  live,
  flakyTest,
  scoped,
  scopedLive,
  layer,
  prop
};
/**
 * @since 1.0.0
 */
export const it = /*#__PURE__*/Object.assign(V.it, {
  ...methods,
  scopedFixtures: /*#__PURE__*/V.it.scoped.bind(V.it)
});
/**
 * @since 1.0.0
 */
export const makeMethods = internal.makeMethods;
/**
 * @since 1.0.0
 */
export const describeWrapped = internal.describeWrapped;
