// @effect/vitest/dist/esm/utils.js
import * as Either from "effect/Either";
import * as Equal from "effect/Equal";
import * as Exit from "effect/Exit";
import * as Option from "effect/Option";
import * as Predicate from "effect/Predicate";
import * as assert from "node:assert";
import { assert as vassert } from "vitest";
// ----------------------------
// Primitives
// ----------------------------
/**
 * Throws an `AssertionError` with the provided error message.
 *
 * @since 0.21.0
 */
export function fail(message) {
  assert.fail(message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
export function deepStrictEqual(actual, expected, message, ..._) {
  assert.deepStrictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is not equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
export function notDeepStrictEqual(actual, expected, message, ..._) {
  assert.notDeepStrictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
export function strictEqual(actual, expected, message, ..._) {
  assert.strictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
export function assertEquals(actual, expected, message, ..._) {
  if (!Equal.equals(actual, expected)) {
    deepStrictEqual(actual, expected, message); // show diff
    fail(message ?? "Expected values to be Equal.equals");
  }
}
/**
 * Asserts that `thunk` does not throw an error.
 *
 * @since 0.21.0
 */
export function doesNotThrow(thunk, message, ..._) {
  assert.doesNotThrow(thunk, message);
}
// ----------------------------
// Derived
// ----------------------------
/**
 * Asserts that `value` is an instance of `constructor`.
 *
 * @since 0.21.0
 */
export function assertInstanceOf(value, constructor, message, ..._) {
  // @ts-ignore
  vassert.instanceOf(value, constructor, message);
}
/**
 * Asserts that `self` is `true`.
 *
 * @since 0.21.0
 */
export function assertTrue(self, message, ..._) {
  strictEqual(self, true, message);
}
/**
 * Asserts that `self` is `false`.
 *
 * @since 0.21.0
 */
export function assertFalse(self, message, ..._) {
  strictEqual(self, false, message);
}
/**
 * Asserts that `actual` includes `expected`.
 *
 * @since 0.21.0
 */
export function assertInclude(actual, expected, ..._) {
  if (Predicate.isString(expected)) {
    if (!actual?.includes(expected)) {
      fail(`Expected\n\n${actual}\n\nto include\n\n${expected}`);
    }
  }
}
/**
 * Asserts that `actual` matches `regexp`.
 *
 * @since 0.21.0
 */
export function assertMatch(actual, regexp, ..._) {
  if (!regexp.test(actual)) {
    fail(`Expected\n\n${actual}\n\nto match\n\n${regexp}`);
  }
}
/**
 * Asserts that `thunk` throws an error.
 *
 * @since 0.21.0
 */
export function throws(thunk, error, ..._) {
  try {
    thunk();
    fail("Expected to throw an error");
  } catch (e) {
    if (error !== undefined) {
      if (Predicate.isFunction(error)) {
        error(e);
      } else {
        deepStrictEqual(e, error);
      }
    }
  }
}
/**
 * Asserts that `thunk` throws an error.
 *
 * @since 0.21.0
 */
export async function throwsAsync(thunk, error, ..._) {
  try {
    await thunk();
    fail("Expected to throw an error");
  } catch (e) {
    if (error !== undefined) {
      if (Predicate.isFunction(error)) {
        error(e);
      } else {
        deepStrictEqual(e, error);
      }
    }
  }
}
// ----------------------------
// Option
// ----------------------------
/**
 * Asserts that `option` is `None`.
 *
 * @since 0.21.0
 */
export function assertNone(option, ..._) {
  deepStrictEqual(option, Option.none());
}
/**
 * Asserts that `option` is `Some`.
 *
 * @since 0.21.0
 */
export function assertSome(option, expected, ..._) {
  deepStrictEqual(option, Option.some(expected));
}
// ----------------------------
// Either
// ----------------------------
/**
 * Asserts that `either` is `Left`.
 *
 * @since 0.21.0
 */
export function assertLeft(either, expected, ..._) {
  deepStrictEqual(either, Either.left(expected));
}
/**
 * Asserts that `either` is `Right`.
 *
 * @since 0.21.0
 */
export function assertRight(either, expected, ..._) {
  deepStrictEqual(either, Either.right(expected));
}
// ----------------------------
// Exit
// ----------------------------
/**
 * Asserts that `exit` is a failure.
 *
 * @since 0.21.0
 */
export function assertFailure(exit, expected, ..._) {
  deepStrictEqual(exit, Exit.failCause(expected));
}
/**
 * Asserts that `exit` is a success.
 *
 * @since 0.21.0
 */
export function assertSuccess(exit, expected, ..._) {
  deepStrictEqual(exit, Exit.succeed(expected));
}
