// @effect/vitest/dist/cjs/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertEquals = assertEquals;
exports.assertFailure = assertFailure;
exports.assertFalse = assertFalse;
exports.assertInclude = assertInclude;
exports.assertInstanceOf = assertInstanceOf;
exports.assertLeft = assertLeft;
exports.assertMatch = assertMatch;
exports.assertNone = assertNone;
exports.assertRight = assertRight;
exports.assertSome = assertSome;
exports.assertSuccess = assertSuccess;
exports.assertTrue = assertTrue;
exports.deepStrictEqual = deepStrictEqual;
exports.doesNotThrow = doesNotThrow;
exports.fail = fail;
exports.notDeepStrictEqual = notDeepStrictEqual;
exports.strictEqual = strictEqual;
exports.throws = throws;
exports.throwsAsync = throwsAsync;
var Either = _interopRequireWildcard(require("effect/Either"));
var Equal = _interopRequireWildcard(require("effect/Equal"));
var Exit = _interopRequireWildcard(require("effect/Exit"));
var Option = _interopRequireWildcard(require("effect/Option"));
var Predicate = _interopRequireWildcard(require("effect/Predicate"));
var assert = _interopRequireWildcard(require("node:assert"));
var _vitest = require("vitest");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
// ----------------------------
// Primitives
// ----------------------------
/**
 * Throws an `AssertionError` with the provided error message.
 *
 * @since 0.21.0
 */
function fail(message) {
  assert.fail(message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
function deepStrictEqual(actual, expected, message, ..._) {
  assert.deepStrictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is not equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
function notDeepStrictEqual(actual, expected, message, ..._) {
  assert.notDeepStrictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
function strictEqual(actual, expected, message, ..._) {
  assert.strictEqual(actual, expected, message);
}
/**
 * Asserts that `actual` is equal to `expected` using the `Equal.equals` trait.
 *
 * @since 0.21.0
 */
function assertEquals(actual, expected, message, ..._) {
  if (!Equal.equals(actual, expected)) {
    deepStrictEqual(actual, expected, message); // show diff
    fail(message ?? "Expected values to be Equal.equals");
  }
}
/**
 * Asserts that `thunk` does not throw an error.
 *
 * @since 0.21.0
 */
function doesNotThrow(thunk, message, ..._) {
  assert.doesNotThrow(thunk, message);
}
// ----------------------------
// Derived
// ----------------------------
/**
 * Asserts that `value` is an instance of `constructor`.
 *
 * @since 0.21.0
 */
function assertInstanceOf(value, constructor, message, ..._) {
  // @ts-ignore
  _vitest.assert.instanceOf(value, constructor, message);
}
/**
 * Asserts that `self` is `true`.
 *
 * @since 0.21.0
 */
function assertTrue(self, message, ..._) {
  strictEqual(self, true, message);
}
/**
 * Asserts that `self` is `false`.
 *
 * @since 0.21.0
 */
function assertFalse(self, message, ..._) {
  strictEqual(self, false, message);
}
/**
 * Asserts that `actual` includes `expected`.
 *
 * @since 0.21.0
 */
function assertInclude(actual, expected, ..._) {
  if (Predicate.isString(expected)) {
    if (!actual?.includes(expected)) {
      fail(`Expected\n\n${actual}\n\nto include\n\n${expected}`);
    }
  }
}
/**
 * Asserts that `actual` matches `regexp`.
 *
 * @since 0.21.0
 */
function assertMatch(actual, regexp, ..._) {
  if (!regexp.test(actual)) {
    fail(`Expected\n\n${actual}\n\nto match\n\n${regexp}`);
  }
}
/**
 * Asserts that `thunk` throws an error.
 *
 * @since 0.21.0
 */
function throws(thunk, error, ..._) {
  try {
    thunk();
    fail("Expected to throw an error");
  } catch (e) {
    if (error !== undefined) {
      if (Predicate.isFunction(error)) {
        error(e);
      } else {
        deepStrictEqual(e, error);
      }
    }
  }
}
/**
 * Asserts that `thunk` throws an error.
 *
 * @since 0.21.0
 */
async function throwsAsync(thunk, error, ..._) {
  try {
    await thunk();
    fail("Expected to throw an error");
  } catch (e) {
    if (error !== undefined) {
      if (Predicate.isFunction(error)) {
        error(e);
      } else {
        deepStrictEqual(e, error);
      }
    }
  }
}
// ----------------------------
// Option
// ----------------------------
/**
 * Asserts that `option` is `None`.
 *
 * @since 0.21.0
 */
function assertNone(option, ..._) {
  deepStrictEqual(option, Option.none());
}
/**
 * Asserts that `option` is `Some`.
 *
 * @since 0.21.0
 */
function assertSome(option, expected, ..._) {
  deepStrictEqual(option, Option.some(expected));
}
// ----------------------------
// Either
// ----------------------------
/**
 * Asserts that `either` is `Left`.
 *
 * @since 0.21.0
 */
function assertLeft(either, expected, ..._) {
  deepStrictEqual(either, Either.left(expected));
}
/**
 * Asserts that `either` is `Right`.
 *
 * @since 0.21.0
 */
function assertRight(either, expected, ..._) {
  deepStrictEqual(either, Either.right(expected));
}
// ----------------------------
// Exit
// ----------------------------
/**
 * Asserts that `exit` is a failure.
 *
 * @since 0.21.0
 */
function assertFailure(exit, expected, ..._) {
  deepStrictEqual(exit, Exit.failCause(expected));
}
/**
 * Asserts that `exit` is a success.
 *
 * @since 0.21.0
 */
function assertSuccess(exit, expected, ..._) {
  deepStrictEqual(exit, Exit.succeed(expected));
}
